### PR TITLE
Make `pipeline` able to load `processor`

### PR DESCRIPTION
**Pull request by qubvel**
# What does this PR do?

Add `processor` to the `pipeline` and refactor tests a bit to support it. Related to 
 - #32490
 - #32059 

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/pipelines/__init__.py
Comment (qubvel) [inline src/transformers/pipelines/__init__.py:939]: For backward compatibility, we can control with `Pipeline` class if we need to load specific processors/tokenizers. For example, for zero-shot object detection, we will need to load only the `processor`, and do not need to load `image_processor` and tokenizer separately. Other legacy pipelines might load tokenizer and image_processor, even if they have processor class.

File: src/transformers/pipelines/base.py
Comment (qubvel) [inline src/transformers/pipelines/base.py:827]: granular control for loading, see comment in the code

File: tests/models/mt5/test_modeling_mt5.py
Comment (qubvel) [inline tests/models/mt5/test_modeling_mt5.py:1084]: This one is skipped on the main but was not skipped on this branch, so I added a skip rule here

File: tests/pipelines/test_pipelines_text_generation.py
Comment (qubvel) [inline tests/pipelines/test_pipelines_text_generation.py:431]: Llama no longer raises exception for long generation, so I added it to exclude list in the test. This test was skipped on main.

File: tests/test_pipeline_mixin.py
Comment (qubvel) [inline tests/test_pipeline_mixin.py:181]: Previously, tokenizer, image processor and feature extractor class names were taken from JSON file of tiny models, however, there is no `*Processor` class there. Now MAPPING lists are used to get class names.

File: tests/test_pipeline_mixin.py
Comment (qubvel) [inline tests/test_pipeline_mixin.py:340]: The order of loading is a bit changed in this test, now the model is loaded first to check if it exists, otherwise, the test is skipped.

**Review by qubvel — COMMENTED** [review]
The main changes for pipeline code:
 - Added loading of `processor` (previously only tokenizer, image_processor and feature_extractor)
 - Added Pipeline class args to control loading
 
 The main changes 
 - Refactored test signatures to support `processor` pass
 - Changed the way how class names are obtained for processors of the testing tiny models in a pipeline. Instead of getting them from JSON file, the MAPPINGs are used.
 
 I highlighted these changes below, please see the comments
 
cc @ydshieh for initial review

**Comment by qubvel** [comment]
@ydshieh yes, you are right, the [following PR](https://github.com/huggingface/transformers/pull/32490) will add processor to the ZeroShotObjectDetection pipeline

**Comment by ydshieh** [comment]
Hi @qubvel I have a general question. The process class could load also a single tokenizer, image processor or feature extractor (if not all components are on a Hub repository). For example,

```python
from transformers import AutoProcessor
AutoProcessor.from_pretrained("bert-base-uncased")
```
will turn out to be a tokenizer instead of an instance of `ProcessorMixin`.

Therefore, in `src/transformers/pipelines/__init__.py`, this part

```python
       # Instantiate processor if needed
        if isinstance(processor, (str, tuple)):
            processor = AutoProcessor.from_pretrained(processor, _from_pipeline=task, **hub_kwargs, **model_kwargs)
```
should have some guard to make sure we really get a `ProcessorMixin` instance.

WDYT?

I will share my thoughts regarding the testing part this afternoon, somehow related to the same consideration.

File: tests/test_pipeline_mixin.py
Comment (ydshieh) [inline tests/test_pipeline_mixin.py:364]: we should make sure `processor` is an instance of `ProcessorMixin`, otherwise reset it to `None`

**Review by ydshieh — COMMENTED** [review]
Some comments :-)

**Comment by qubvel** [comment]
Screenshot 2024-08-14 at 16 51 57
do not have a `Reply...` field regarding MAPPING (github bug?)

I did it because in the original tiny-models JSON file we do not have separate lists for image processors and feature extractors, I found they are stored in the same list (even processors are sometimes in this list).

File: tests/test_pipeline_mixin.py
Comment (qubvel) [reply tests/test_pipeline_mixin.py:396]: can you provide more details, please? this function is used only here and not using image_processor or feature_extractor (I updated the function signature accordingly)

File: tests/test_pipeline_mixin.py
Comment (qubvel) [reply tests/test_pipeline_mixin.py:364]: should we? Class is not checked for feature extractor or image processor, I guess we might want to load anything that might be loaded with AutoProcessor class while it keeps the pipeline working.

File: tests/test_pipeline_mixin.py
Comment (qubvel) [reply tests/test_pipeline_mixin.py:181]: I did it because in the original tiny-models JSON file we do not have separate lists for image processors and feature extractors, I found they are stored in the same list (even processors are sometimes in this list).

File: tests/test_pipeline_mixin.py
Comment (ydshieh) [reply tests/test_pipeline_mixin.py:396]: Yeah, it's more a design instead of what is actually happening so far.

The picture is that, a pipeline test is involved pipeline class (task), model, tokenizer/image processor/feature extractor/processor, and all of these **could** determine what we want to perform. (although currently only model and tokenizer appears in the body)

File: tests/models/mt5/test_modeling_mt5.py
Comment (qubvel) [reply tests/models/mt5/test_modeling_mt5.py:1084]: Yes, this test was not running on the main branch. One of the problem of current tests: we are testing a set of cases, and if one of the cases (e.g. the first one) falls under skipTest(...) the rest test cases are not running. Probably it's better to return case status, and then aggregate statuses in the main test function running on multiple cases.

File: tests/test_pipeline_mixin.py
Comment (ydshieh) [reply tests/test_pipeline_mixin.py:364]: > Class is not checked for feature extractor or image processor

This is because they are not a composite, and are not design to fallback to load partially.

For example, this works
```
from transformers import AutoProcessor
AutoProcessor.from_pretrained("bert-base-uncased")
```
but this fails
```
from transformers import AutoImageProcessor
AutoImageProcessor.from_pretrained("bert-base-uncased")
```
and same
```
from transformers import AutoTokenizer
AutoTokenizer.from_pretrained("google/vit-base-patch16-224")
```

File: tests/models/mt5/test_modeling_mt5.py
Comment (qubvel) [reply tests/models/mt5/test_modeling_mt5.py:1084]: A simple example to reproduce:

```python
import unittest

class TestCases(unittest.TestCase):

    def test_multiple_cases(self):
        for i in range(10):
            self.run_test(i)

    def run_test(self, i):
        if i == 0:
            self.skipTest("Skip this test")
        elif i == 5:
            raise ValueError("This test failed")
```

**Comment by ydshieh** [comment]
> I did it because in the original tiny-models JSON file we do not have separate lists for image processors and feature extractors, I found they are stored in the same list (even processors are sometimes in this list).

Yes, you are correct. It is because I treated them (image processor / feature exactor) equally, and later I decided to make `processor` join into this pipeline testing

I see better now why you need this change. Could you, at least for now, simply use the information from the summary file `processor_classes` field, and by looking at training part to decide to assign them to one of the  `tokenizer_names`, `feature_extractor_names`, `processor_names` etc.?
``

File: tests/test_pipeline_mixin.py
Comment (qubvel) [reply tests/test_pipeline_mixin.py:384]: It is designed the same way on the `main` now, according to the comment some instances might not be loaded due to optional dependencies
https://github.com/huggingface/transformers/blob/af638c4afea81ac251f1ae63497789be79e0baaa/tests/test_pipeline_mixin.py#L256-L267

File: tests/test_pipeline_mixin.py
Comment (qubvel) [reply tests/test_pipeline_mixin.py:402]: Thanks for the clarification, I understand your point. 
While this is a design decision, I'm not entirely comfortable with it. The function isn't widely used or overridden in subclasses, it's only used once in the codebase. Not using arguments might cause confusion about what's actually happening in the function and could give the false impression that the image processor and tokenizer are being validated by it, but they are not. We can always add additional arguments to the function if they become necessary.

File: tests/test_pipeline_mixin.py
Comment (qubvel) [reply tests/test_pipeline_mixin.py:370]: Got it, addressed in https://github.com/huggingface/transformers/pull/32514/commits/2f561479290970fa8beaf668b504df39a2608f4f

**Comment by qubvel** [comment]
@ydshieh Thanks for reviewing, I addressed the comments, please have a look!

> Once we start to have the code using processor, and if we support it for some currently existing pipelines, it should support:
> - either not use the processor at all (i.e. currently implementation): for backward compatibility
> - or completely use the processor only (but need to verify the results are the same)

This is what I tried to achieve with granular control for each pipeline class by adding `_load_*` attributes 
https://github.com/huggingface/transformers/pull/32514#discussion_r1711361484

**Comment by qubvel** [comment]
@ydshieh could you please have a look once again?

File: tests/test_pipeline_mixin.py
Comment (ydshieh) [reply tests/test_pipeline_mixin.py:377]: Currently on `main`, this `run_pipeline_test` only take a single `tokenizer_name` and a **single** `processor_name`, and try to load them. So if a specified `processor_name` could not be loaded, we can skip it.

However, here in this PR, we try to **collect** the processors and put them into the dictionary `processors`. So logically it's better not to skip the test during this collection.

(But in practice, this may not produce any difference)

File: tests/test_pipeline_mixin.py
Comment (ydshieh) [inline tests/test_pipeline_mixin.py:327]: This looks like not the correct docstring?

File: tests/test_pipeline_mixin.py
Comment (ydshieh) [reply tests/test_pipeline_mixin.py:402]: works for me.

(FYI: unused arguments exist in `class Pipeline.__init__` and I agree with your above point)

File: tests/test_pipeline_mixin.py
Comment (ydshieh) [inline tests/test_pipeline_mixin.py:188]: the first condition is not necessary at this point.

File: tests/test_pipeline_mixin.py
Comment (ydshieh) [inline tests/test_pipeline_mixin.py:170]: if we don't add `None` later in this method (while it's empty), then in `run_model_pipeline_tests`, 

> test_cases = [ ....]

will be empty if `tokenizer_names` is empty or `image_processor_names` is empty or ... (etc).

But we should allow the test case to exist even if only one of them present: for example, a text only model or a vision only model.

**Review by ydshieh — COMMENTED** [review]
Just a few nits but overall good!

Let me know if my comments are clear 🙏

**Comment by yonigozlan** [comment]
Hi @qubvel @ydshieh ! Just wondering if anything is blocking this PR from getting merged, as it is needed for the image-text-to-text pipeline. Thanks!

**Comment by qubvel** [comment]
Hi @yonigozlan, just a lack of time 🙂 I will address comments this week, and hopefully it can be merged then!

File: tests/test_pipeline_mixin.py
Comment (qubvel) [reply tests/test_pipeline_mixin.py:188]: otherwise we can get an error while doing `tiny_model_summary[model_arch_name]`

File: tests/test_pipeline_mixin.py
Comment (qubvel) [reply tests/test_pipeline_mixin.py]: addressed in https://github.com/huggingface/transformers/pull/32514/commits/403cd0b810413da5e462f28410c224e9d8eca39d

**Comment by qubvel** [comment]
@Rocketknight1 can you please review the PR, if that aligns with plans for `pipeline` refactoring/standardization

**Comment by Rocketknight1** [comment]
I'm sorry, I missed that ping somehow! Reviewing now.

File: src/transformers/pipelines/__init__.py
Comment (Rocketknight1) [inline src/transformers/pipelines/__init__.py:939]: I think this conditional is redundant - if `feature_extractor` is a `str` then it will always be `not None` as well.

File: src/transformers/pipelines/__init__.py
Comment (Rocketknight1) [inline src/transformers/pipelines/__init__.py:944]: Same here!

File: src/transformers/pipelines/__init__.py
Comment (Rocketknight1) [inline src/transformers/pipelines/__init__.py:946]: And here!

File: tests/test_pipeline_mixin.py
Comment (Rocketknight1) [inline tests/test_pipeline_mixin.py:285]: Note for other reviewers: When a model doesn't support one of these preprocessor types, the list of names will be `[None]`, **not** an empty list. When I first looked at this quadruple-loop I was worried that `test_cases` would be empty if any of the four name lists was empty, but that should never occur. For safety, we might want to add a check for that, though, since it depends on behaviour in a different function that might be modified later!

**Review by Rocketknight1 — APPROVED** [review]
Just finished the review, with a few comments.

Overall summary for core maintainers:
- This is a nice PR! It's clean, and although it seems very large, almost all of the changes are boilerplate adding kwargs and fixing spelling mistakes in a lot of files.
- The actual changes you need to look at are `__init__.py`, `base.py` and `test_pipeline_mixin.py`
- I made some comments, but they're nits and shouldn't affect any of the PR logic.

cc @lysandrejik for core maintainer review!

File: src/transformers/pipelines/__init__.py
Comment (qubvel) [reply src/transformers/pipelines/__init__.py]: addressed in https://github.com/huggingface/transformers/pull/32514/commits/d553dab8a33c4144b29c649115fb3a41cd92d6bf

File: src/transformers/pipelines/__init__.py
Comment (qubvel) [reply src/transformers/pipelines/__init__.py]: addressed in https://github.com/huggingface/transformers/pull/32514/commits/d553dab8a33c4144b29c649115fb3a41cd92d6bf

File: src/transformers/pipelines/__init__.py
Comment (qubvel) [reply src/transformers/pipelines/__init__.py]: addressed in https://github.com/huggingface/transformers/pull/32514/commits/d553dab8a33c4144b29c649115fb3a41cd92d6bf

File: tests/test_pipeline_mixin.py
Comment (qubvel) [reply tests/test_pipeline_mixin.py:288]: You are right, let's have this check closer to the loop to ensure we always have something to test
https://github.com/huggingface/transformers/pull/32514/commits/6ff4e68ec9caa17f4f5f2535f00f0fbca5fda115

Comment (ydshieh) [reply tests/test_pipeline_mixin.py:288]: After this change, you can also remove

```
            if not isinstance(processor_names, (list, tuple)):
                processor_names = [processor_names]
```
(which was there for the same reason)

Comment (qubvel) [reply tests/test_pipeline_mixin.py:288]: I left it there to make sure we do not have `str` processor_names

**Review by LysandreJik — COMMENTED** [review]
This looks good! I'm mostly concerned about having good and clear documentation and removing unnecessary warnings.

Pipelines are the very first transformers object many users interact with -> ensuring that they're aware of the info they need, without being hammered by warnigns should be an absolute goal of ours.

Thanks for working on this important change!

File: src/transformers/pipelines/__init__.py
Comment (qubvel) [reply src/transformers/pipelines/__init__.py:668]: I added a separate Note section to highlight that we should not provide all types of `processors` at once
https://github.com/huggingface/transformers/pull/32514/commits/e7127176ae83dd401a0e0358aa6cecd5db80fe9b and refer to a specific pipeline in case one would like to provide them explicitly.

For each specific pipeline we have only required processors args in docs section configured with docs decorator. e.g. here
https://github.com/huggingface/transformers/blob/e782e95e3465b66a377c0a7fe95f8f10cd87459d/src/transformers/pipelines/image_classification.py#L53

File: src/transformers/pipelines/base.py
Comment (qubvel) [reply src/transformers/pipelines/base.py]: Tried to make it more clear https://github.com/huggingface/transformers/pull/32514/commits/9793e012a62dfda59942b586e0e21b612b5e0cfb

File: src/transformers/pipelines/__init__.py
Comment (qubvel) [reply src/transformers/pipelines/__init__.py]: Thanks for the questions!  
There was a discussion [here](https://github.com/huggingface/transformers/pull/32514#discussion_r1717215386). The main idea is that `AutoProcessor` can load not only processors but also basic processing classes, like `tokenizer`. 

Indeed, a misconfiguration in the model-pipeline-processor setup could trigger this and raising a warning + dropping the `processor` might be sufficient **only when the `processor` isn't needed in the `pipeline` at all.** 

However, with granular control, such a case shouldn't occur. Therefore, it seems more appropriate to replace the warning with an error to clearly indicate a misconfiguration. Otherwise, the error will happen later with a less clear message because the `processor` is None.

Comment (qubvel) [reply src/transformers/pipelines/__init__.py]: Changed to error raise here https://github.com/huggingface/transformers/pull/32514/commits/bcce4dceb19b39225b11acbcc495a754fa6c802d

File: src/transformers/pipelines/__init__.py
Comment (qubvel) [reply src/transformers/pipelines/__init__.py:676]: Updated in https://github.com/huggingface/transformers/pull/32514/commits/699669503629a834013d9c2107396051b3b9327e

**Comment by qubvel** [comment]
@LysandreJik thanks for reviewing, I addressed the comment: made clearer docs and removed the warning. Please have a look whenever you have bandwidth.

P.S. The test failure looks unrelated.

**Review by LysandreJik — APPROVED** [review]
Ok great, thanks for the changes @qubvel!

**Pull request merged**
